File: backend/app/schemas/application_automation.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


AutomationStatus = Literal[
    "draft",
    "awaiting_login",
    "awaiting_confirmation",
    "submitting",
    "submitted",
    "verification_required",
    "failed",
    "cancelled",
]


class AutomationPlatformInfo(BaseModel):
    key: str
    name: str
    description: str
    login_url: str
    capabilities: list[str]
    browser_bridge_available: bool
    daily_submission_limit: int
    cooldown_seconds: int


class LoginStatus(BaseModel):
    platform_key: str
    status: Literal["logged_in", "logged_out", "unknown", "bridge_unavailable"]
    target_id: str | None = None
    evidence: list[str] = []
    message: str


class AutomationTaskCreate(BaseModel):
    platform_key: Literal["zhaopin"]
    job_id: int = Field(gt=0)
    resume_id: int = Field(gt=0)
    greeting_id: int | None = Field(default=None, gt=0)
    greeting_content: str | None = Field(default=None, max_length=1000)


class AutomationTaskRead(BaseModel):
    id: int
    platform_key: str
    job_id: int
    resume_id: int
    greeting_id: int | None
    application_record_id: int | None
    status: AutomationStatus
    job_title_snapshot: str
    company_snapshot: str
    job_url_snapshot: str
    resume_title_snapshot: str
    greeting_snapshot: str
    confirmation_token: str
    preview: dict[str, object]
    external_result: dict[str, object]
    confirmed_by_user: bool
    error_message: str
    prepared_at: datetime | None
    submitted_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AutomationSubmitRequest(BaseModel):
    confirmed: bool
    confirmation_token: str = Field(min_length=8, max_length=80)


class AutoApplyPlanRequest(BaseModel):
    greeting_content: str | None = Field(default=None, max_length=1000)


class AutoApplyCandidateRead(BaseModel):
    job_id: int
    title: str
    company_name: str
    city: str
    salary_min: int | None
    salary_max: int | None
    source_url: str
    score: int
    reasons: list[str]
    blockers: list[str]
    eligible: bool
    task_id: int | None = None
    task_status: AutomationStatus | None = None


class AutoApplyPlanRead(BaseModel):
    platform_key: Literal["zhaopin"]
    resume_id: int
    resume_title: str
    salary_floor: int
    eligible_count: int
    queued_count: int
    skipped_count: int
    candidates: list[AutoApplyCandidateRead]
    message: str


class AutomationCancelRequest(BaseModel):
    reason: str = Field(default="用户选择不投递", max_length=200)


class ZhaopinSearchImportRequest(BaseModel):
    keyword: str | None = Field(default=None, max_length=120)
    city: str | None = Field(default=None, max_length=80)
    page_limit: int = Field(default=3, ge=1, le=5)
    import_limit: int = Field(default=20, ge=1, le=50)


class ZhaopinSearchExclusionClearRequest(BaseModel):
    search_signature: str = Field(min_length=64, max_length=64, pattern=r"^[0-9a-f]{64}$")


class ZhaopinSearchExclusionClearRead(BaseModel):
    deleted_count: int
    message: str


class ZhaopinSearchCandidateRead(BaseModel):
    title: str
    company_name: str
    city: str
    salary_text: str
    salary_min: int | None
    salary_max: int | None
    degree_required: str
    experience_required: str
    company_size: str
    description_loaded: bool
    auto_blacklisted: bool = False
    source_url: str
    eligible: bool
    score: int
    reasons: list[str]
    blockers: list[str]
    job_id: int | None = None
    import_action: Literal["created", "updated"] | None = None


class ZhaopinSearchImportRead(BaseModel):
    query: str
    search_url: str
    pages_requested: int
    scanned_count: int
    eligible_count: int
    created_count: int
    updated_count: int
    auto_blacklisted_count: int
    history_skipped_count: int = 0
    search_signature: str
    candidates: list[ZhaopinSearchCandidateRead]
    message: str


class AutoApplyRunRequest(ZhaopinSearchImportRequest):
    greeting_content: str | None = Field(default=None, max_length=1000)


class AutoApplyRunRead(BaseModel):
    search: ZhaopinSearchImportRead
    plan: AutoApplyPlanRead
    message: str

File: frontend/src/app/lib/applicationAutomation.ts
export type TaskStatus =
  | "draft"
  | "awaiting_login"
  | "awaiting_confirmation"
  | "submitting"
  | "submitted"
  | "verification_required"
  | "failed"
  | "cancelled";

export type AutoApplyCandidate = {
  job_id: number;
  title: string;
  company_name: string;
  city: string;
  salary_min: number | null;
  salary_max: number | null;
  source_url: string;
  score: number;
  reasons: string[];
  blockers: string[];
  eligible: boolean;
  task_id: number | null;
  task_status: TaskStatus | null;
};

export type AutoApplyPlan = {
  platform_key: "zhaopin";
  resume_id: number;
  resume_title: string;
  salary_floor: number;
  eligible_count: number;
  queued_count: number;
  skipped_count: number;
  candidates: AutoApplyCandidate[];
  message: string;
};

export type ZhaopinSearchImport = {
  scanned_count: number;
  eligible_count: number;
  created_count: number;
  updated_count: number;
  auto_blacklisted_count: number;
  message: string;
};

export type AutoApplyRun = {
  search: ZhaopinSearchImport;
  plan: AutoApplyPlan;
  message: string;
};

export type AutoApplyRunRequest = {
  page_limit: number;
  import_limit: number;
  greeting_content: string | null;
};

export const HARD_MINIMUM_SALARY = 3000;

export function resolveSalaryFloor(profile: { salary_min: number | null } | null | undefined) {
  return Math.max(HARD_MINIMUM_SALARY, profile?.salary_min ?? HARD_MINIMUM_SALARY);
}

export function selectBatchPreparationCandidates(
  candidates: AutoApplyCandidate[],
  limit = 5,
) {
  return candidates
    .filter((candidate) => candidate.eligible && candidate.task_id !== null && candidate.task_status !== "cancelled")
    .slice(0, limit);
}

export function candidateActionLabel(status: TaskStatus | null) {
  return status === "awaiting_confirmation" ? "确认投递" : "预览并确认";
}

type Fetcher = (input: RequestInfo | URL, init?: RequestInit) => Promise<Response>;

export async function runAutoApply(
  apiBase: string,
  request: AutoApplyRunRequest,
  fetcher: Fetcher = fetch,
): Promise<AutoApplyRun> {
  const response = await fetcher(
    `${apiBase}/api/application-automation/auto-apply/run`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(request),
    },
  );
  if (!response.ok) {
    const body = await response.json().catch(() => null) as { detail?: string } | null;
    throw new Error(body?.detail ?? "智联搜索和待投递清单生成失败");
  }
  return response.json() as Promise<AutoApplyRun>;
}
